### PR TITLE
Add screenshot test for highlights visibility when layer selection/active layer changes

### LINKED ISSUE #7315: Add screenshot test for hiding highlights of unselected layers
## 🧰 Task

Copied from: https://github.com/napari/napari/pull/7289#issuecomment-2388072768

typically I make a screenshot check with something that won't be too sensitive to rendering details. See for example [my test in 6654](https://github.com/napari/napari/pull/6654/files#diff-b34aee7c8386aa63aaa1809904227439f5f8ffb2469dc2939532564fdbaf78d6R573-R578):

```
    screenshot = viewer.screenshot(canvas_only=True, flash=False)
    # center should have a hole
    # use -1 because edge is exactly halfway down.
    center = (screenshot.shape[0] // 2 - 1, screenshot.shape[1] // 2)
    # ignore alpha channel
    np.testing.assert_array_equal(screenshot[center][:3], 0)
```

So in this case, I would check that the highlight color is or isn't present in the screenshot, based on the active layer.

### PR DESCRIPTION
**Pull request by dalthviz**
# References and relevant issues

Closes #7315 

# Description

Add a test checking the highlight behavior when changing selection of a layer (Shapes layer). Check that the select layer data highlight is visible only when the layer is selected/active

### CODE REVIEW

File: napari/_tests/test_with_screenshot.py [hunk 1]
```diff
@@ -545,3 +545,39 @@ def test_blending_modes_with_canvas(make_napari_viewer):
...
+
+
+@skip_local_popups
+def test_active_layer_highlight_visibility(make_napari_viewer):
+    viewer = make_napari_viewer(show=True)
+    viewer.window._qt_viewer.set_welcome_visible(False)
+
+    # take initial screenshot (full black/empty screenshot since welcome message is hidden)
+    launch_screenshot = viewer.screenshot(canvas_only=True, flash=False)
+    assert launch_screenshot[..., :-1].max() == 0
```
Comment (DragaDoncila) [inline napari/_tests/test_with_screenshot.py:565]: How come we're not just asserting the whole array max is 0 here?

File: napari/_tests/test_with_screenshot.py [hunk 2]
```diff
@@ -545,3 +545,39 @@ def test_blending_modes_with_canvas(make_napari_viewer):
...
+    assert highlight_screenshot[..., :-1].max() > 0
+
+    # clear viewer layer selection
+    viewer.layers.selection.clear()
+
+    # there shouldn't be a highlight so a new screenshot shouldn't have something visible
+    no_highlight_screenshot = viewer.screenshot(canvas_only=True, flash=False)
+    assert no_highlight_screenshot[..., :-1].max() == 0
+
+    # selecte again the layer with the rectangle shape
```
Comment (DragaDoncila) [inline napari/_tests/test_with_screenshot.py]:
```suggestion
    # select again the layer with the rectangle shape
```

**Review by DragaDoncila — APPROVED** [review]
Looks ok to me, just a question and a nitpick!

File: napari/_tests/test_with_screenshot.py (hunk 1 above)
Comment (dalthviz) [reply napari/_tests/test_with_screenshot.py:565]: I think the last index of each element in the array represents its alpha value of 255 (1 when normalized to a 0 to 1 scale). So if I change the asserts here to be something like:

```python
assert launch_screenshot.max() == 0
```

The tests will fail and as error you would see something like the following:

```python-traceback
>       assert launch_screenshot.max() == 0
E       assert 255 == 0
E        +  where 255 = <built-in method max of numpy.ndarray object at 0x0000029298073270>()
E        +    where <built-in method max of numpy.ndarray object at 0x0000029298073270> = array([[[  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        ...,\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255]],\n\n       [[  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        ...,\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255]],\n\n       [[  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        ...,\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255]],\n\n       ...,\n\n       [[  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        ...,\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255]],\n\n       [[  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        ...,\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255]],\n\n       [[  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        ...,\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255],\n        [  0,   0,   0, 255]]], dtype=uint8).max

napari\_tests\test_with_screenshot.py:557: AssertionError
```
However maybe changing the array slicing syntax and adding a comment, so something like:

```python
# check screenshot ignoring alpha
assert launch_screenshot[..., :3].max() == 0
```

Could make things more clear? Let me know!

Comment (DragaDoncila) [reply napari/_tests/test_with_screenshot.py:565]: ohhh no of course that's fine, perhaps just add a comment saying that we're ignoring the alpha value!

**Review by Czaki — APPROVED** [review]

**Review by Czaki — COMMENTED** [review]
Could you update this test to use `qt_viewer` fixture?
https://github.com/napari/docs/pull/505

The `QtViewer` also has a method `screenshot` that could be used in place of `viewer.screenshot`

File: napari/_tests/test_with_screenshot.py [hunk 3]
```diff
@@ -10,6 +10,14 @@
 )
 
 
+@pytest.fixture
+def qt_viewer(qt_viewer_):
+    # show the qt_viewer and hide its welcome widget
+    qt_viewer_.show()
+    qt_viewer_.set_welcome_visible(False)
+    return qt_viewer_
```
Comment (Czaki) [inline napari/_tests/test_with_screenshot.py:18]:
```suggestion
    yield qt_viewer_
    qt_viewer_.hide()
```

Comment (dalthviz) [reply napari/_tests/test_with_screenshot.py:18]: @Czaki seems like with this change the test started raising a `RuntimeError` with PyQt6. Checking the `qt_viewer_` fixture, seems like a cleaning function is being passed to the `qtbot.addWidget` call so maybe doing the `yield` + `hide()` call is not necessary? 🤔 :

https://github.com/napari/napari/blob/b9dc8dca1e1afa7567586e7ffa7a36cc7cd9fb6f/napari/conftest.py#L402

Let me know!

Comment (Czaki) [reply napari/_tests/test_with_screenshot.py:18]: You are correct. May you put proper comment in https://github.com/napari/docs/pull/505 to clarify that hide is not neccesary?

Comment (dalthviz) [reply napari/_tests/test_with_screenshot.py:18]: Seems like https://github.com/napari/docs/pull/505 has been already merged but could it be worthy then to create a new PR adding around https://github.com/napari/docs/blob/b9e26a385af6c728ca621f42e0292524281a9737/docs/developers/contributing/testing.md?plain=1#L290 something like:

```
including hidding and clearing any references to viewer instances via a cleaning function passed to the `qtbot.addWidget` as the `before_close_func` kwarg
```

For the moment will revert the suggestion but let me know if that change/addition over the docs makes sense!

**Comment by DragaDoncila** [comment]
@Czaki what's the status on this PR now, are we happy with the changes?

**Comment by Czaki** [comment]
@DragaDoncila code look good. I do not have time to test it manually yet.

**Comment by DragaDoncila** [comment]
>  I do not have time to test it manually yet.

You mean like, running the test manually? The original functionality for ensuring highlight isn't visible when layer isn't active has already been merged

**Review by Czaki — APPROVED** [review]

**Pull request merged**